sql/idempotent/005-vectorizer.sql

```diff
@@ -0,0 +1,1178 @@
+            ;
+            if not _found then
+                raise exception 'pg_cron extension not found';
+            end if;
+            _extension_schema = 'cron';
+        when _implementation = 'none' then
+            return null;
+        else
+            raise exception 'scheduling implementation not recognized';
+    end case;
+
+    -- schedule the job using the implementation chosen
+    case _implementation
+        when 'pg_cron' then
+            -- schedule the work proc with pg_cron
+            select pg_catalog.format
+            ( $$select %I.schedule(%L, %L, $sql$call ai._vectorizer_async_ext_job(null, pg_catalog.jsonb_build_object('vectorizer_id', %s))$sql$)$$
+            , _extension_schema
+            , pg_catalog.concat('vectorizer_', vectorizer_id)
+            , pg_catalog.jsonb_extract_path_text(scheduling, 'schedule')
+            , vectorizer_id
+            ) into strict _sql
+            ;
+            execute _sql into strict _job_id;
+        when 'timescaledb' then
+            -- schedule the work proc with timescaledb background jobs
+            select pg_catalog.format
+            ( $$select %I.add_job('ai._vectorizer_async_ext_job'::regproc, %s, config=>%L)$$
+            , _extension_schema
+            , ( -- gather up the arguments
+                select string_agg
+                ( case s.key
+                    when 'schedule_interval' then pg_catalog.format('%L', s.value)
+                    else pg_catalog.format('%s=>%L', s.key, s.value)
+                  end
+                , ', '
+                order by x.ord
+                )
+                from pg_catalog.jsonb_each_text(scheduling) s
+                inner join
+                unnest(array['schedule_interval', 'initial_start', 'fixed_schedule', 'timezone']) with ordinality x(key, ord)
+                on (s.key = x.key)
+              )
+            , pg_catalog.jsonb_build_object('vectorizer_id', vectorizer_id)::text
+            ) into strict _sql
+            ;
+            execute _sql into strict _job_id;
+    end case;
+    return _job_id;
+end
+$func$
+language plpgsql volatile security invoker
+set search_path to pg_catalog, pg_temp
+;
+
+-------------------------------------------------------------------------------
+-- create_vectorizer
+create or replace function ai.create_vectorizer
+( source regclass
+, embedding jsonb
+, chunking jsonb
+, indexing jsonb default ai.indexing_diskann()
+, formatting jsonb default ai.formatting_python_template()
+, scheduling jsonb default ai.scheduling_timescaledb()
+, asynchronous bool default true -- TODO: remove?
+, external bool default true -- TODO: remove?
+, target_schema name default null
+, target_table name default null
+, view_schema name default null
+, view_name name default null
+, queue_schema name default null
+, queue_table name default null
+, grant_to name[] default array['vectorizer'] -- TODO: what is the final role name we want to use here?
```

Comment:
I think worker_roles or embedding_worker_roles is a better name as it make it more clear what these grants do
```suggestion
, worker_roles name[] default array['vectorizer'] -- TODO: what is the final role name we want to use here?
```

Reply:
It's not restricted to workers though. The caller can add arbitrary users to the array to grant them usage of all the parts created.